lib/spack/spack/test/llnl/util/tty/log.py

```diff
@@ -0,0 +1,472 @@
+    ctl.status()
+    ctl.wait_disabled_fg()
+
+    ctl.bg()             # send to background
+    ctl.tstp()           # immediately stop
+    ctl.status()
+    ctl.wait_stopped()
+    ctl.wait_disabled()
+
+    ctl.cont()           # continue in bg, wait a bit
+    ctl.status()
+    ctl.wait_running()
+
+    ctl.fg()             # bring to foreground
+    ctl.status()
+    ctl.wait_disabled_fg()
+
+    time.sleep(.1)       # allow some time to NOT print
+    ctl.write(b'v\n')    # enable verbose output
+    ctl.status()
+    time.sleep(.1)       # allow time to print
+
+    ctl.bg()             # send to background (keep running)
+    ctl.status()
+    ctl.wait_disabled()
+
+    ctl.tstp()           # stop in bg
+    ctl.status()
+    ctl.wait_stopped()
+
+    ctl.cont()           # continue in bg for a bit
+    ctl.status()
+    ctl.wait_running()
+
+    ctl.fg()             # bring to foreground
+    ctl.status()
+    ctl.wait_disabled_fg()
+
+    ctl.write(b'v\n')    # disable verbose and wait a bit
+    ctl.status()
+
+    ctl.bg()             # back to background
+    ctl.status()
+    ctl.wait_disabled()
+
+
+def mock_shell_integration_1(proc, ctl, attrs):
+    _mock_shell_integration(proc, ctl, attrs)
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_integration_loop(proc, ctl, attrs):
+    for i in range(10):
+        _mock_shell_integration(proc, ctl, attrs)
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_integration_1_no_termios(proc, ctl, attrs):
+    _mock_shell_integration_no_termios(proc, ctl, attrs)
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_integration_loop_no_termios(proc, ctl, attrs):
+    for i in range(10):
+        _mock_shell_integration_no_termios(proc, ctl, attrs)
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+@contextlib.contextmanager
+def no_termios():
+    saved = llnl.util.tty.log.termios
+    llnl.util.tty.log.termios = None
+    yield
+    llnl.util.tty.log.termios = saved
+
+
+@contextlib.contextmanager
+def nullcontext():
+    yield
+
+
+@pytest.mark.skipif(not which("ps"), reason="requires ps utility")
+@pytest.mark.skipif(not termios, reason="requires termios support")
+@pytest.mark.parametrize('test_fn,termios_on_or_off', [
+    # tests with termios
+    (mock_shell_fg, nullcontext),
+    (mock_shell_bg, nullcontext),
+    (mock_shell_bg_fg, nullcontext),
+    (mock_shell_fg_bg, nullcontext),
+    (mock_shell_tstp_cont, nullcontext),
+    (mock_shell_tstp_tstp_cont, nullcontext),
+    (mock_shell_tstp_tstp_cont_cont, nullcontext),
+    # tests without termios
+    (mock_shell_fg_no_termios, no_termios),
+    (mock_shell_bg, no_termios),
+    (mock_shell_bg_fg_no_termios, no_termios),
+    (mock_shell_fg_bg_no_termios, no_termios),
+    (mock_shell_tstp_cont, no_termios),
+    (mock_shell_tstp_tstp_cont, no_termios),
+    (mock_shell_tstp_tstp_cont_cont, no_termios),
+])
+def test_foreground_background_simple(test_fn, termios_on_or_off):
```

Comment:
By "simple" I think we mean "doesn't fail" - I'd prefer the name be updated. I think it's also worth explaining in a docstring that this is to check for issues like `SIGTOU` raising an error for the running process.

Reply:
done